src/uct/ugni/smsg/ugni_smsg_ep.c

```diff
@@ -174,6 +174,15 @@ uct_ugni_smsg_ep_am_common_send(uct_ugni_smsg_ep_t *ep, uct_ugni_smsg_iface_t *i
 {
     gni_return_t gni_rc;
 
+    if(iface->super.send_q_len > 0 && iface->super.send_in_pending == 0) {
+        /* if other jobs are ahead of this one, try sending those first */
+        iface->super.send_in_pending = 1;
+        uint32_t attempts = ucs_min(iface->config.smsg_max_credit, iface->super.send_q_len);
+        ucs_arbiter_dispatch(&iface->super.arbiter, attempts,
+                             uct_ugni_ep_process_pending, NULL);
+        iface->super.send_in_pending = 0;
```

Comment:
It seems better to attempt a dispatch, see if that clears things up enough to allow the send to happen, and if not let the upper layer decide what to do.